src/test/java/tests/PageHeaderIteratorTests.java

```diff
@@ -0,0 +1,131 @@
+package tests;
+
+import github.GitHubClientEx;
+import github.update.PageHeaderIterator;
+import org.eclipse.egit.github.core.Milestone;
+import org.eclipse.egit.github.core.client.PagedRequest;
+import org.junit.Test;
+import org.mockserver.client.server.MockServerClient;
+import org.mockserver.integration.ClientAndServer;
+import org.mockserver.model.Header;
+import org.mockserver.model.HttpRequest;
+import org.mockserver.model.Parameter;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+import static org.eclipse.egit.github.core.client.IGitHubConstants.CONTENT_TYPE_JSON;
+import static org.eclipse.egit.github.core.client.IGitHubConstants.SEGMENT_PULLS;
+import static org.eclipse.egit.github.core.client.IGitHubConstants.SEGMENT_REPOS;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.mockserver.model.HttpRequest.request;
+import static org.mockserver.model.HttpResponse.response;
+
+public class PageHeaderIteratorTests {
+    /**
+     * Tests that PageHeaderIterator throws a NoSuchElement exception when its next
+     * method is called with a initial request to an non-existent repository
+     * @throws NoSuchElementException
+     */
+    @Test(expected = NoSuchElementException.class)
+    public void testHeaderIteratorInvalidRepo() throws NoSuchElementException {
+        GitHubClientEx client = new GitHubClientEx();
+
+        Map<String, String> params = new HashMap<>();
+        params.put("state", "all");
+
+        PagedRequest<Milestone> request = new PagedRequest<>();
+        String path = SEGMENT_REPOS + "/nonexistentrepo";
+        request.setUri(path);
+        request.setResponseContentType(CONTENT_TYPE_JSON);
+        request.setParams(params);
+
+        PageHeaderIterator iter = new PageHeaderIterator(request, client, "ETag");
+        if (iter.hasNext()) {
+            iter.next();
+        }
+    }
+
+    /**
+     * Tests that a PageHeaderIterator correctly retrieves ETag headers for 3 pages from
+     * a mocked server that conform to GitHub API's pagination specifications and terminates afterwards.
+     * @throws NoSuchElementException
+     */
+    @Test
+    public void testHeaderIterator() throws NoSuchElementException, IOException {
+        MockServerClient mockServer = ClientAndServer.startClientAndServer(8888);
+
+        HttpRequest page1Request = createMockServerPagedHeaderRequest("1");
+        String page1Etag = "aaf65fc6b10d5afbdc9cd0aa6e6ada4c";
+        String page1Link =
+             "<https://localhost:8888/repositories/20919534/pulls?state=all&per_page=100&page=2>; rel=\"next\", "
```

Comment:
Yup, we can abstract this here. It's only for the tests in this particular case though since the `repositories/*` part varies for each set of pagination results. Most of the time, we will use separate files to store responses/headers from GitHub API so we shouldn't have to do this too often.